include/ensmallen_bits/nelder_mead/nelder_mead_impl.hpp

```diff
@@ -0,0 +1,241 @@
+    delta(delta),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    simplexer(simplexer),
+    adaptiveDefault(false)
+{ /* Nothing to do. */ }
+
+//! Optimize the function.
+template<typename SimplexerType>
+template<typename ArbitraryFunctionType,
+         typename MatType,
+         typename... CallbackTypes>
+typename MatType::elem_type NelderMeadType<SimplexerType>::Optimize(
+    ArbitraryFunctionType& function,
+    MatType& iterate,
+    CallbackTypes&&... callbacks)
+{
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+  typedef typename MatTypeTraits<MatType>::BaseMatType BaseMatType;
+
+  // Make sure that we have the methods that we need.  Long name...
+  traits::CheckArbitraryFunctionTypeAPI<ArbitraryFunctionType,
+      BaseMatType>();
+  RequireDenseFloatingPointType<BaseMatType>();
+
+  // Controls early termination of the optimization process.
+  bool terminate = false;
+
+  // Controls the shrink step.
+  bool shrink = false;
+
+  // Problem dimensionality.
+  const size_t dim = iterate.n_rows;
+
+  // Use default adaptive parameters scheme as used in "Implementing the
+  // Nelder-Mead simplex algorithm with adaptive parameters" by F. Gao and L.
+  // Han (2010).
+  if (adaptiveDefault)
+  {
+    alpha = 1;
+    beta = 1.0 + 2.0 / (double) dim;
+    gamma = 0.75 - 0.5 / (double) dim;
+    delta = 1.0 - 1.0 / (double) dim;
+  }
+
+  // Construct the initial simplex.
+  MatType simplex;
+  simplexer.Simplex(simplex, iterate, function);
+
+  arma::rowvec fx(dim + 1);
+  for (size_t i = 0; i < dim + 1; ++i)
+  {
+    fx(i) = function.Evaluate(simplex.col(i));
+    Callback::Evaluate(*this, function, simplex.col(i), fx(i), callbacks...);
+  }
+
+  // Get the indices that correspond to the ordering of the function values
+  // at the vertices. order(0) is the index in the simplex of the vertex
+  // with the lowest function value, and order(dim) is the index in the
+  // simplex of the vertex with the highest function value.
+  arma::uvec order = arma::sort_index(fx);
+
+  terminate |= Callback::BeginOptimization(*this, function, iterate,
+      callbacks...);
+  for (size_t i = 1; i != maxIterations && !terminate; ++i)
+  {
+    const MatType M = arma::mean(simplex.cols(order.subvec(0, dim - 1)), 1);
+    const MatType xref = (1.0 + alpha) * M - alpha * simplex.col(order(dim));
+    const ElemType fref = function.Evaluate(xref);
+    Callback::Evaluate(*this, function, xref, fref, callbacks...);
+
+    if (fref < fx(order(0)))
+    {
+      // Compute expansion.
+      const MatType xexp = (1.0 + alpha * beta) * M - alpha * beta *
+          simplex.col(order(dim));
+      const ElemType fexp = function.Evaluate(xexp);
+      Callback::Evaluate(*this, function, xexp, fexp, callbacks...);
+
+      // Shift values by one to the right and swap the last with the first.
+      const size_t fLowest = order(dim);
+      for (size_t d = dim; d > 0; --d)
+        order(d) = order(d - 1);
```

Comment:
Here too you could use `subvec()` to copy all at once, or, if you use a max-heap this step may not be needed.  (I believe `std::priority_queue` is a max-heap.)